app/assets/javascripts/discourse/app/services/composer.js

```diff
@@ -1498,36 +1499,36 @@ export default class ComposerController extends Controller {
 
     return new Promise((resolve) => {
       if (this.get("model.hasMetaData") || this.get("model.replyDirty")) {
-        const modal = showModal("discard-draft", {
-          model: this.model,
-          modalClass: "discard-draft-modal",
-        });
         const overridesDraft =
           this.model.composeState === Composer.OPEN &&
           this.model.draftKey === opts.draftKey &&
           [Composer.EDIT_SHARED_DRAFT, Composer.EDIT].includes(opts.action);
         const showSaveDraftButton = this.model.canSaveDraft && !overridesDraft;
-        modal.setProperties({
-          showSaveDraftButton,
-          onDestroyDraft: () => {
-            return this.destroyDraft()
-              .then(() => {
+        getOwner(this)
+          .lookup("service:modal")
+          .show(DiscardDraftModal, {
```

Comment:
Since this code is running in the context of an Ember Service, I think it would be cleaner to 'inject' the modal service onto the Composer Service, and then we can reference it like `this.modal.show(DiscardDraftModal, ...)`.

(there are some other injections at the top of the `class ComposerService` definition)